File: intersection-of-two-arrays/Solution349.java
// https://leetcode.com/problems/intersection-of-two-arrays/

import java.util.*;

class Solution349 {

    // T: O(n+m), S: O(n+m)
    public int[] intersection(int[] nums1, int[] nums2) {
        var seen = new HashSet<Integer>();
        for (int n : nums1) {
            seen.add(n);
        }
        var list = new ArrayList<Integer>();
        for (int n : nums2) {
            if (seen.contains(n)) {
                list.add(n);
                seen.remove(n);
            }
        }
        var result = new int[list.size()];
        for (int i = 0; i < list.size(); i++) {
            result[i] = list.get(i);
        }
        return result;
    }

    public static void main(String[] args) {
        var sol = new Solution349();
        var nums1 = new int[] { 1, 2, 2, 1 };
        var nums2 = new int[] { 4, 9, 5 };
        System.out.println(Arrays.toString(sol.intersection(nums1, nums2))); // [2]
        nums1 = new int[] { 4, 9, 5 };
        nums2 = new int[] { 9, 4, 9, 8, 4 };
        System.out.println(Arrays.toString(sol.intersection(nums1, nums2))); // [9,4]
    }

}

File: intersection-of-two-arrays/solution349.go
// https://leetcode.com/problems/intersection-of-two-arrays/

package main

import "fmt"

// T: O(n+m), S: O(n+m)
func intersection(nums1 []int, nums2 []int) []int {
	seen := make(map[int]bool)
	result := make([]int, 0)
	for _, n := range nums1 {
		seen[n] = true
	}
	for _, n := range nums2 {
		if seen[n] {
			result = append(result, n)
			seen[n] = false
		}
	}
	return result
}

func main() {
	nums1 := []int{1, 2, 2, 1}
	nums2 := []int{4, 9, 5}
	fmt.Println(intersection(nums1, nums2)) // [2]
	nums1 = []int{4, 9, 5}
	nums2 = []int{9, 4, 9, 8, 4}
	fmt.Println(intersection(nums1, nums2)) // [9,4]
}
